File: src/main/java/ru/arc/util/GuiUtils.java
package ru.arc.util;

import ru.arc.ARC;
import com.github.stefvanschie.inventoryframework.gui.GuiItem;
import com.github.stefvanschie.inventoryframework.gui.type.ChestGui;
import lombok.extern.slf4j.Slf4j;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver;
import org.bukkit.Material;
import org.bukkit.entity.HumanEntity;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.scheduler.BukkitTask;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.function.Supplier;

import static ru.arc.util.Logging.error;
import static ru.arc.util.TextUtil.strip;

@Slf4j
public class GuiUtils {

    private static Map<BgKey, GuiItem> backgrounds = new HashMap<>();

    record BgKey(Material material, int model) {
    }

    public static GuiItem background(Material material, int model) {
        GuiItem guiItem = backgrounds.get(new BgKey(material, model));
        if (guiItem != null) return guiItem;

        ItemStack bgItem = new ItemStack(material);
        ItemMeta meta = bgItem.getItemMeta();
        if (model != 0) meta.setCustomModelData(model);
        meta.displayName(Component.text(" "));
        bgItem.setItemMeta(meta);
        guiItem = new GuiItem(bgItem, inventoryClickEvent -> inventoryClickEvent.setCancelled(true));
        backgrounds.put(new BgKey(material, model), guiItem);

        return guiItem;
    }

    public static BukkitTask temporaryChange(ItemStack stack, Component display, List<Component> lore, long ticks, Runnable callback, TagResolver resolver) {
        ItemMeta meta = stack.getItemMeta();
        final Component oldDisplay = meta.displayName();
        final List<Component> oldLore = meta.lore();
        if (display != null) meta.displayName(strip(display));
        if (lore != null && !lore.isEmpty()) meta.lore(lore.stream().map(TextUtil::strip).toList());
        stack.setItemMeta(meta);
        if (ticks < 0) return null;
        return new BukkitRunnable() {
            @Override
            public void run() {
                ItemMeta meta1 = stack.getItemMeta();
                meta1.displayName(strip(oldDisplay));
                if (oldLore != null) meta1.lore(oldLore.stream().map(TextUtil::strip).toList());
                else meta1.lore(null);
                stack.setItemMeta(meta1);
                callback.run();
            }
        }.runTaskLater(ARC.plugin, ticks);
    }

    public static BukkitTask temporaryChange(ItemStack stack, Component display, List<Component> lore, long ticks, Runnable callback) {
        return temporaryChange(stack, display, lore, ticks, callback, null);
    }

    public static boolean cooldownCheck(GuiItem guiItem, UUID playerUuid, ChestGui chestGui) {
        long cooldown = CooldownManager.cooldown(playerUuid, "gui_click");
        if (cooldown > 0) {
            temporaryChange(guiItem.getItem(),
                    strip(Component.text("Не кликайте так быстро!", NamedTextColor.RED)),
                    null, cooldown, () -> {
                        if (chestGui != null) chestGui.update();
                    }
            );
            return false;
        }
        CooldownManager.addCooldown(playerUuid, "gui_click", 10);
        return true;
    }

    public static void constructAndShowAsync(Supplier<ChestGui> supplier, HumanEntity player, int delay) {
        CompletableFuture.supplyAsync(() -> {
                    try {
                        return supplier.get();
                    } catch (Exception e) {
                        error("Error opening menu", e);
                        return null;
                    }
                })
                .thenAccept(gui -> new BukkitRunnable() {
                    @Override
                    public void run() {
                        if (gui == null) {
                            error("Gui is null {}", supplier);
                            return;
                        }
                        gui.show(player);
                    }
                }.runTaskLater(ARC.plugin, delay));
    }

    public static void constructAndShowAsync(Supplier<ChestGui> supplier, HumanEntity player) {
        constructAndShowAsync(supplier, player, 3);
    }

    public static GuiItem background(Material material) {
        return background(material, 0);
    }

    public static GuiItem background() {
        return background(Material.GRAY_STAINED_GLASS_PANE, 11000);
    }


}

File: src/main/kotlin/ru/arc/util/GuiUtils.kt
package ru.arc.util

import com.github.stefvanschie.inventoryframework.gui.GuiItem
import com.github.stefvanschie.inventoryframework.gui.type.ChestGui
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver
import org.bukkit.Material
import org.bukkit.entity.HumanEntity
import org.bukkit.inventory.ItemStack
import ru.arc.core.ScheduledTask
import ru.arc.core.async
import ru.arc.core.delayed
import ru.arc.core.ticks
import ru.arc.util.Logging.error
import ru.arc.util.TextUtil.strip
import java.util.UUID
import java.util.function.Supplier

object GuiUtils {

    private val backgrounds = mutableMapOf<BgKey, GuiItem>()

    data class BgKey(val material: Material, val model: Int)

    /**
     * Clears the backgrounds cache. Used in tests to reset state.
     */
    @JvmStatic
    fun clearBackgrounds() {
        backgrounds.clear()
    }

    @JvmStatic
    fun background(material: Material, model: Int): GuiItem {
        val key = BgKey(material, model)
        return backgrounds.getOrPut(key) {
            guiItem(material) {
                display(" ")
                if (model != 0) modelData(model)
                onClick { it.isCancelled = true }
            }
        }
    }

    /**
     * Temporarily change an item's display and lore, then restore after delay.
     * Uses Task DSL for scheduling.
     */
    @JvmStatic
    fun temporaryChange(
        stack: ItemStack,
        display: Component?,
        lore: List<Component>?,
        ticks: Long,
        callback: Runnable,
        resolver: TagResolver?,
    ): ScheduledTask? {
        val meta = stack.itemMeta
        val oldDisplay = meta?.displayName()
        val oldLore = meta?.lore()

        if (display != null) meta?.displayName(strip(display))
        if (lore != null && lore.isNotEmpty()) {
            meta?.lore(lore.map { strip(it)!! })
        }
        stack.itemMeta = meta

        if (ticks < 0) return null

        return delayed(ticks.ticks) {
            val meta1 = stack.itemMeta
            meta1?.displayName(strip(oldDisplay))
            if (oldLore != null) {
                meta1?.lore(oldLore.map { strip(it)!! })
            } else {
                meta1?.lore(null)
            }
            stack.itemMeta = meta1
            callback.run()
        }
    }

    @JvmStatic
    fun temporaryChange(
        stack: ItemStack,
        display: Component?,
        lore: List<Component>?,
        ticks: Long,
        callback: Runnable,
    ): ScheduledTask? = temporaryChange(stack, display, lore, ticks, callback, null)

    @JvmStatic
    fun cooldownCheck(guiItem: GuiItem, playerUuid: UUID, chestGui: ChestGui?): Boolean {
        val cooldown = CooldownManager.cooldown(playerUuid, "gui_click")
        if (cooldown > 0) {
            temporaryChange(
                guiItem.item,
                strip(Component.text("Не кликайте так быстро!", NamedTextColor.RED)),
                null,
                cooldown,
                Runnable { chestGui?.update() },
            )
            return false
        }
        CooldownManager.addCooldown(playerUuid, "gui_click", 10)
        return true
    }

    /**
     * Construct and show a GUI on the main server thread after an optional delay.
     * ChestGui construction requires the main thread (Bukkit API restriction),
     * so this always runs the supplier synchronously on the main thread.
     */
    @JvmStatic
    fun constructAndShowAsync(supplier: Supplier<ChestGui>, player: HumanEntity, delay: Int) {
        delayed(delay.ticks) {
            val gui =
                try {
                    supplier.get()
                } catch (e: Exception) {
                    error("Error constructing GUI", e)
                    return@delayed
                }
            gui.show(player)
        }
    }

    @JvmStatic
    fun constructAndShowAsync(supplier: Supplier<ChestGui>, player: HumanEntity) {
        constructAndShowAsync(supplier, player, 3)
    }

    @JvmStatic
    fun background(material: Material): GuiItem {
        return background(material, 0)
    }

    @JvmStatic
    fun background(): GuiItem {
        return background(Material.GRAY_STAINED_GLASS_PANE, 11000)
    }
}

